lib/dice/src/cert.rs

```diff
@@ -29,121 +31,97 @@ pub trait Cert {
     fn as_bytes(&self) -> &[u8];
     fn as_mut_bytes(&mut self) -> &mut [u8];
 
-    const SERIAL_NUMBER_START: usize;
-    const SERIAL_NUMBER_END: usize;
-
-    fn get_serial_number(&self) -> u8 {
-        u8::from_be_bytes(
-            self.as_bytes()[Self::SERIAL_NUMBER_START..Self::SERIAL_NUMBER_END]
-                .try_into()
-                .unwrap_lite(),
-        )
-    }
-
-    fn set_serial_number(mut self, sn: u8) -> Self
+    fn read_range<'a, T>(&'a self, r: Range<usize>) -> T
     where
-        Self: Sized,
+        T: TryFrom<&'a [u8]>,
```

Comment:
There's nothing inherent about TryFrom<&[u8]> that says it can only return Ok if the entire slice is used.  It happens that you use a pattern of `let x: [u8; X] = self.read_range(R)` which means T is TryFrom<&'a [u8]> for [u8; N] which does fail if the slice and array lengths are different (I had to go look that up: https://doc.rust-lang.org/src/core/array/mod.rs.html#177-186).  I'm not sure if that was an intentional choice here or whether T should always consume the entire slice provided.  This feels like an easy place to make a mistake but I can't think of any easy way to detect such an error.

Reply:
This pattern was intentional. It was the best option I could come up with that didn't rely on an unstable feature (generic_const_exprs). If you're OK with enabling incomplete / unstable features this can be updated, or we can wait for the feature to stabilize. I've created an issue to track this: https://github.com/oxidecomputer/hubris/issues/778